### ISSUE TITLE
hasBlock always returns true

### ISSUE THREAD

**Issue by simonihmig** [embroider-build/embroider#373]
I encountered a weird issue when trying to test building the addon `ember-bootstrap` with embroider: where components are used in non-block form (https://github.com/kaliber5/ember-bootstrap/blob/master/addon/templates/components/common/bs-modal-simple.hbs#L28-L34), the `hasBlock` helper inside that component (https://github.com/kaliber5/ember-bootstrap/blob/master/addon/templates/components/common/bs-modal/footer.hbs#L1) always returns true, though false is expected. This could be confirmed with adding a `{{log hasBlock}}`: while the normal build returns false, just changing the build to use embroider makes this return true! 🤔

The best reproduction I have is this WIP PR to add an ember-try scenario to run the test suite with an embroider build: https://github.com/kaliber5/ember-bootstrap/pull/930. The 10 failing tests (all other pass) for the embroider build (https://travis-ci.org/kaliber5/ember-bootstrap/jobs/633547428?utm_medium=notification&utm_source=github_status) all indicate to be related to this `hasBlock` behaviour. 

This can be reproduced locally by running `FASTBOOT_DISABLED=true ember t` on that branch (needs some hacks to disable FastBoot due to embroider's existing issues)

Furthermore I digged into the compiled bundles, and found that the template that does the non-block form invocation (https://github.com/kaliber5/ember-bootstrap/blob/master/addon/templates/components/common/bs-modal-simple.hbs#L28-L34) is compiled differently to Glimmer Opcodes (just quoting the modal/footer component invocation):

Normal build:

```
[5,\"bs-modal/footer\",[],[[\"@closeTitle\",\"@submitTitle\",\"@submitButtonType\",\"@onClose\",\"@onSubmit\"],[[23,0,[\"closeTitle\"]],[23,6,[]],[23,0,[\"submitButtonType\"]],[23,0,[\"close\"]],[23,0,[\"doSubmit\"]]]]],
[0,\"\\n    \"]],\"parameters\":[]}],[0,\"\\n\\n    \"],
```

Embroider build:

```
[5,\"bs-modal/footer\",[],[[\"@closeTitle\",\"@submitTitle\",\"@submitButtonType\",\"@onClose\",\"@onSubmit\"],[[23,0,[\"closeTitle\"]],[23,6,[]],[23,0,[\"submitButtonType\"]],[23,0,[\"close\"]],[23,0,[\"doSubmit\"]]]],
{\"statements\":[],\"parameters\":[]}],[0,\"\\n    \"]],\"parameters\":[]}],[0,\"\\n\\n    \"],
```

Not sure how this wire format needs to be interpreted exactly, but there is an obvious difference (`statements`) that could eventually suggest that the component *is* wrongly called in block-form in the embroider build (thus `hasBlock` is not to blame)!? 🤔

**Comment by ef4** [comment]
Thanks for reporting this, I haven't had a chance to dig in yet but it certainly sounds weird!

**Comment by simonihmig** [comment]
For the record, just files this issue: https://github.com/glimmerjs/glimmer-vm/issues/1025

The side effect (wrong `hasBlock`) is very similar, but not sure if the root issue is really the same or even related...

**Comment by simonihmig** [comment]
Seems the underlying issue has been fixed (https://github.com/glimmerjs/glimmer-vm/issues/1025#issuecomment-589978015), I cannot observe this anymore. 🥳

**Comment by ef4** [comment]
After hitting this issue in an app, we discovered the root cause of this problem is https://github.com/glimmerjs/glimmer-vm/pull/941

It only occurs if you use `hasBlock` in an addon (not an app) and if that addon uses some custom AST transforms.

The glimmer-vm bug was fixed in ember-source 3.17, which is therefore our minimum safe version to avoid this bug.
